lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -937,13 +1043,31 @@ void SILGlobalOpt::optimizeGlobalAccess(SILGlobalVariable *SILG,
 
 }
 
-bool SILGlobalOpt::run() {
-  for (auto &F : *Module) {
+void SILGlobalOpt::reset() {
+  AllocGlobalStore.clear();
+  GlobalVarStore.clear();
+  GlobalAddrMap.clear();
+  GlobalAccessMap.clear();
+  GlobalLoadMap.clear();
+  GlobalInitCallMap.clear();
+}
 
-    // Don't optimize functions that are marked with the opt.never attribute.
-    if (!F.shouldOptimize())
-      continue;
+void SILGlobalOpt::collectUsesOfInstructionForDeletion(SILInstruction *inst) {
```

Comment:
There might be a utility for doing this already. The safest way is to use a worklist, instead of a recursive function, in case you come across cycles or deep paths in the def-use graph.

Reply:
Good to know, I'll use a worklist.

Reply:
I updated the code to use a worklist but I don't think it works well here. We can't iterate over a worklist which is necessary in `tryRemoveGlobalAddr`. I'm going to change it back to a `SmallVector` and use `eraseUsesOfInstruction` when removing instructions.